common/src/Microsoft.Azure.IIoT.Core/src/Http/HttpClient.cs

```diff
@@ -105,38 +105,60 @@ public sealed class HttpClient : IHttpClient {
-                            ContentHeaders = response.Content.Headers,
-                            Content = await response.Content.ReadAsByteArrayAsync()
-                        };
-                        if (result.IsError()) {
-                            _logger.Warning("{method} to {uri} returned {code} (took {elapsed}).",
-                                httpMethod, httpRequest.Uri, response.StatusCode, sw.Elapsed,
-                                 result.GetContentAsString(Encoding.UTF8));
+
+                using (var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(ct)) {
+                    try {
+                        wrapper.Request.Method = httpMethod;
+                        using (var response = await client.SendAsync(wrapper.Request, linkedCts.Token)) {
+                            var result = new HttpResponse {
+                                ResourceId = httpRequest.ResourceId,
+                                StatusCode = response.StatusCode,
+                                Headers = response.Headers,
+                                ContentHeaders = response.Content.Headers,
+                                Content = await response.Content.ReadAsByteArrayAsync()
+                            };
+                            if (result.IsError()) {
+                                _logger.Warning("{method} to {uri} returned {code} (took {elapsed}).",
+                                    httpMethod, httpRequest.Uri, response.StatusCode, sw.Elapsed,
+                                     result.GetContentAsString(Encoding.UTF8));
+                            }
+                            else {
+                                _logger.Verbose("{method} to {uri} returned {code} (took {elapsed}).",
+                                    httpMethod, httpRequest.Uri, response.StatusCode, sw.Elapsed);
+                            }
+                            return result;
                         }
-                        else {
-                            _logger.Verbose("{method} to {uri} returned {code} (took {elapsed}).",
-                                httpMethod, httpRequest.Uri, response.StatusCode, sw.Elapsed);
+                    }
+                    catch (HttpRequestException e) {
+                        var errorMessage = e.Message;
+                        if (e.InnerException != null) {
+                            errorMessage += " - " + e.InnerException.Message;
                         }
-                        return result;
+                        _logger.Warning("{method} to {uri} failed (after {elapsed}) : {message}!",
+                            httpMethod, httpRequest.Uri, sw.Elapsed, errorMessage);
+                        _logger.Verbose(e, "{method} to {uri} failed (after {elapsed}) : {message}!",
+                            httpMethod, httpRequest.Uri, sw.Elapsed, errorMessage);
+                        throw new HttpRequestException(errorMessage, e);
```

Comment:
i spot some duplicate code here :-)

Reply:
I was not able to use exception filters here because of having two distinct execution flows in case of `OperationCanceledException`. Instead, I've defines a local `generateHttpRequestException()` function to do the formatting and return `HttpRequestException`.

Reply:
ok, it works...

---
In reply to: [494957320](https://github.com/Azure/Industrial-IoT/pull/731#discussion_r494957320) [](ancestors = 494957320)